File: scripts/lib/promax_line_classifier.py
#!/usr/bin/env python3
"""Line-level ad vs enhancement classification for PROMAX merge."""

from __future__ import annotations

import re
from typing import Literal, Optional

LineVerdict = Literal["ad", "enhance", "neutral"]

# Script-path / rule-name signals
AD_SCRIPT_MARKERS = (
    "adblock",
    "/adblock/",
    ".adblock.",
    "anti-ad",
    "blockad",
    "remove_ads",
    "remove-ads",
    "去广告",
    "/ads/",
    "advertising",
    "reject",
    "spam",
    "tracking",
    "xwebads",
    "adultraplus",
    "allinone",
    "mock",
    "bilibili.helper",
    "helper.v2.js",
    "camoufox",  # generic block lists — rarely unlock
)

ENHANCE_SCRIPT_MARKERS = (
    "/enhanced/",
    ".enhanced.",
    "/enhance/",
    "biliuniverse/enhanced",
    "/global/",
    ".global.",
    "biliuniverse/global",
    "/redirect/",
    ".redirect.",
    "unlock",
    "解锁",
    "unblock",
    "vip",
    "premium",
    "crack",
    "破解",
    "camscanner.js",
    "external_links_unlock",
    "weixin_external",
    "1080p",
    "bilibili_json.js",  # deezertidal quality unlock
    "dualsubs",
    "iringo",
    "weatherkit",
    "zheye",
    "哲也",
    "translation",
    "翻译",
    "nsfw",
    "region",
    "intl",
    "bypass",
    "绕过",
    "pip",
    "画中画",
    "optimize",
    "功能增强",
)

ENHANCE_NAME_PREFIXES = (
    "bilibili.enhanced",
    "bilibili.global",
    "bilibili.redirect",
    "youtube.enhance",
    "wechat_enhance",
    "iringo.",
    "dualsubs",
)

AD_NAME_PREFIXES = (
    "bilibili.adblock",
    "adblock",
    "去广告",
    "xwebads",
    "anti-ad",
)

REJECT_POLICIES = frozenset(
    {
        "REJECT",
        "REJECT-DROP",
        "REJECT-NO-DROP",
        "REJECT-TINYGIF",
        "REJECT-IMG",
    }
)


def _line_lower(line: str) -> str:
    return line.strip().lower()


def _script_path(line: str) -> str:
    m = re.search(r"script-path\s*=\s*([^,\s]+)", line, re.I)
    return (m.group(1) if m else "").lower()


def _entry_name(line: str) -> str:
    if "=" not in line:
        return ""
    return line.split("=", 1)[0].strip().lower()


def classify_promax_line(line: str, section: Optional[str] = None) -> LineVerdict:
    """Classify one non-comment module line for PROMAX ingestion."""
    stripped = line.strip()
    if not stripped or stripped.startswith("#"):
        return "neutral"

    low = _line_lower(stripped)
    sec = (section or "").strip().lower()

    if sec == "mitm":
        if "hostname" not in low:
            return "neutral"
        if any(m in low for m in ("unlock", "解锁", "camscanner", "intsig.net")):
            return "enhance"
        return "neutral"

    if sec == "general":
        return "enhance"

    if sec == "header rewrite":
        return "enhance"

    name = _entry_name(stripped)
    spath = _script_path(stripped)

    if sec == "script" or "script-path=" in low:
        if any(m in name for m in AD_NAME_PREFIXES) or any(
            m in spath for m in AD_SCRIPT_MARKERS
        ):
            return "ad"
        if any(m in name for m in ENHANCE_NAME_PREFIXES) or any(
            m in spath for m in ENHANCE_SCRIPT_MARKERS
        ):
            return "enhance"
        if "helper" in name and "bili" in name:
            return "ad"
        return "enhance"

    if sec == "url rewrite" or (
        stripped.startswith("^") and (" reject" in low or " _ reject" in low)
    ):
        if " reject" in low or " _ reject" in low:
            return "ad"
        return "enhance"

    if sec in ("map local", "body rewrite", "maplocal"):
        if any(m in low for m in ("广告", "adcard", "splash", "deliver", "flash", "e-commerce")):
            return "ad"
        if 'data="{}"' in low or "data='{}'" in low:
            return "ad"
        if "del(.data.payment)" in low or "payment" in low and "del(" in low:
            return "ad"
        return "neutral"

    if sec == "rule" or stripped.split(",", 1)[0].upper() in (
        "DOMAIN",
        "DOMAIN-SUFFIX",
        "DOMAIN-KEYWORD",
        "DOMAIN-REGEX",
        "DOMAIN-WILDCARD",
        "URL-REGEX",
        "IP-CIDR",
        "IP-CIDR6",
        "USER-AGENT",
        "PROCESS-NAME",
        "DEST-PORT",
    ):
        parts = [p.strip() for p in stripped.split(",")]
        policy = parts[-1].upper() if parts else ""
        if policy in REJECT_POLICIES:
            if any(m in low for m in ("unlock", "解锁", "vip", "premium", "crack")):
                return "enhance"
            return "ad"
        # Bare DOMAIN / IP rules (no trailing policy) → merged as REJECT in AdBlock lists
        if parts and parts[0].upper() in (
            "DOMAIN",
            "DOMAIN-SUFFIX",
            "DOMAIN-KEYWORD",
            "DOMAIN-REGEX",
            "DOMAIN-WILDCARD",
            "URL-REGEX",
            "IP-CIDR",
            "IP-CIDR6",
            "USER-AGENT",
            "PROCESS-NAME",
            "DEST-PORT",
        ):
            if any(m in low for m in ("unlock", "解锁", "vip", "premium", "crack")):
                return "enhance"
            return "ad"
        return "enhance"

    if any(m in low for m in ENHANCE_SCRIPT_MARKERS):
        return "enhance"
    if any(m in low for m in AD_SCRIPT_MARKERS):
        return "ad"
    return "neutral"


def should_keep_promax_line(line: str, section: Optional[str] = None) -> bool:
    verdict = classify_promax_line(line, section)
    if verdict == "ad":
        return True
    if verdict == "enhance":
        return False
    sec = (section or "").strip().lower()
    if sec == "mitm" and "hostname" in line.lower():
        return True
    return False


def module_ingest_mode(path: str, text: str) -> str:
    """
    Return ingest mode for a module file:
      - skip: unlock-only or no ad content
      - full: pure ad module (all lines considered)
      - split: mixed ad + enhancement — line-level filter
    """
    import os

    name = os.path.basename(path).lower()
    if "解锁" in name or "unlock" in name:
        if not any(t in name for t in ("去广告", "adblock", "anti-ad")):
            return "skip"

    ad_lines = 0
    enhance_lines = 0
    current_section: Optional[str] = None

    for raw in text.splitlines():
        stripped = raw.strip()
        if stripped.startswith("[") and stripped.endswith("]"):
            current_section = stripped[1:-1]
            continue
        if not stripped or stripped.startswith("#"):
            continue
        v = classify_promax_line(stripped, current_section)
        if v == "ad":
            ad_lines += 1
        elif v == "enhance":
            enhance_lines += 1

    if ad_lines == 0:
        return "skip"
    if enhance_lines > 0:
        return "split"
    return "full"

File: go_scripts/pkg/hub/promax_line_classifier.go
package hub

import (
	"path/filepath"
	"regexp"
	"strings"
)

type LineVerdict string

const (
	VerdictAd      LineVerdict = "ad"
	VerdictEnhance LineVerdict = "enhance"
	VerdictNeutral LineVerdict = "neutral"
)

var AD_SCRIPT_MARKERS = []string{
	"adblock", "/adblock/", ".adblock.", "anti-ad", "blockad", "remove_ads",
	"remove-ads", "_remove_", "/remove", "去广告", "/ads/", "advertising", "reject",
	"spam", "tracking", "xwebads", "adultraplus", "allinone", "mock",
	"bilibili.helper", "helper.v2.js", "camoufox", "rednote", "redpaper",
	"weibo_ads", "weibo_main", "wool_scripts", "kelee.one/resource/javascript",
	"proxyrules/main/script", "ithome_remove", "zhihu_remove", "amap_remove",
	"cainiao_remove", "taobao_remove", "jd_remove", "spotify_remove", "bilicomic_remove",
}

var AD_ENTRY_NAME_MARKERS = []string{
	"去广告", "adblock", "anti-ad", "xwebads", "移除", "精简", "推广", "开屏", "净化", "广告", "killad", "mock",
}

var ENHANCE_SCRIPT_MARKERS = []string{
	"/enhanced/", ".enhanced.", "/enhance/", "biliuniverse/enhanced", "/global/",
	".global.", "biliuniverse/global", "/redirect/", ".redirect.", "unlock", "解锁",
	"unblock", "vip", "premium", "crack", "破解", "camscanner.js", "external_links_unlock",
	"weixin_external", "1080p", "bilibili_json.js", "dualsubs", "iringo", "weatherkit",
	"zheye", "哲也", "translation", "翻译", "nsfw", "region", "intl", "bypass", "绕过",
	"pip", "画中画", "optimize", "功能增强",
}

var ENHANCE_NAME_PREFIXES = []string{
	"bilibili.enhanced", "bilibili.global", "bilibili.redirect", "youtube.enhance",
	"wechat_enhance", "iringo.", "dualsubs",
}

var AD_NAME_PREFIXES = []string{
	"bilibili.adblock", "adblock", "去广告", "xwebads", "anti-ad",
}

var REJECT_POLICIES = map[string]bool{
	"REJECT":         true,
	"REJECT-DROP":    true,
	"REJECT-NO-DROP": true,
	"REJECT-TINYGIF": true,
	"REJECT-IMG":     true,
}

var PURE_AD_MODULE_NAME_TOKENS = []string{
	"去广告", "adblock", "anti-ad", "xwebads", "adultraplus", "allinone", "all-in-one",
	"rednote", "redpaper", "微博去广告", "remove_ads",
}

var scriptPathReStr = regexp.MustCompile(`(?i)script-path\s*=\s*([^,\s]+)`)

func scriptPath(line string) string {
	m := scriptPathReStr.FindStringSubmatch(line)
	if len(m) > 1 {
		return strings.ToLower(m[1])
	}
	return ""
}

func entryName(line string) string {
	if !strings.Contains(line, "=") {
		return ""
	}
	parts := strings.SplitN(line, "=", 2)
	return strings.ToLower(strings.TrimSpace(parts[0]))
}

func containsAny(s string, markers []string) bool {
	for _, m := range markers {
		if strings.Contains(s, m) {
			return true
		}
	}
	return false
}

func SourceIsPureAdModule(sourcePath string) bool {
	if sourcePath == "" {
		return false
	}
	low := strings.ToLower(filepath.Base(sourcePath))
	return containsAny(low, PURE_AD_MODULE_NAME_TOKENS)
}

func scriptIsAd(name string, spath string, low string) bool {
	if containsAny(name, AD_NAME_PREFIXES) || containsAny(name, AD_ENTRY_NAME_MARKERS) {
		if containsAny(name, []string{"解锁", "unlock", "vip", "premium", "破解", "crack"}) {
			return false
		}
		return true
	}
	if containsAny(spath, AD_SCRIPT_MARKERS) {
		return true
	}
	if strings.Contains(name, "helper") && strings.Contains(name, "bili") {
		return true
	}
	return false
}

func scriptIsEnhance(name string, spath string) bool {
	if containsAny(name, ENHANCE_NAME_PREFIXES) || containsAny(spath, ENHANCE_SCRIPT_MARKERS) {
		return true
	}
	if containsAny(name, []string{"解锁", "unlock", "vip", "premium", "破解", "crack"}) {
		return true
	}
	return false
}

func ClassifyPromaxLine(line string, section string, sourcePath string) LineVerdict {
	stripped := strings.TrimSpace(line)
	if stripped == "" || strings.HasPrefix(stripped, "#") {
		return VerdictNeutral
	}

	low := strings.ToLower(stripped)
	sec := strings.ToLower(strings.TrimSpace(section))
	pureAdSource := SourceIsPureAdModule(sourcePath)

	if sec == "mitm" {
		if !strings.Contains(low, "hostname") {
			return VerdictNeutral
		}
		if containsAny(low, []string{"unlock", "解锁", "camscanner", "intsig.net"}) {
			return VerdictEnhance
		}
		return VerdictNeutral
	}

	if sec == "general" || sec == "header rewrite" {
		return VerdictEnhance
	}

	name := entryName(stripped)
	spath := scriptPath(stripped)

	if sec == "script" || strings.Contains(low, "script-path=") {
		if scriptIsEnhance(name, spath) {
			return VerdictEnhance
		}
		if scriptIsAd(name, spath, low) {
			return VerdictAd
		}
		if pureAdSource {
			return VerdictAd
		}
		return VerdictEnhance
	}

	if sec == "url rewrite" || (strings.HasPrefix(stripped, "^") && (strings.Contains(low, " reject") || strings.Contains(low, " _ reject"))) {
		if strings.Contains(low, " reject") || strings.Contains(low, " _ reject") {
			return VerdictAd
		}
		return VerdictEnhance
	}

	if sec == "map local" || sec == "body rewrite" || sec == "maplocal" {
		if strings.Contains(low, "http-response-jq") || strings.Contains(low, "http-request-jq") {
			if containsAny(low, ENHANCE_SCRIPT_MARKERS) {
				return VerdictEnhance
			}
			return VerdictAd
		}
		if strings.HasPrefix(stripped, "http-response ") || strings.HasPrefix(stripped, "http-request ") {
			return VerdictAd
		}
		if strings.HasPrefix(stripped, "^") && (strings.Contains(low, "data-type=") || strings.Contains(low, "data=\"{}\"") || strings.Contains(low, "data='{}'")) {
			return VerdictAd
		}
		if containsAny(low, []string{"广告", "adcard", "splash", "deliver", "flash", "e-commerce"}) {
			return VerdictAd
		}
		if strings.Contains(low, "data=\"{}\"") || strings.Contains(low, "data='{}'") {
			return VerdictAd
		}
		if strings.Contains(low, "del(.data.payment)") || (strings.Contains(low, "payment") && strings.Contains(low, "del(")) {
			return VerdictAd
		}
		if pureAdSource && strings.HasPrefix(stripped, "^") {
			return VerdictAd
		}
		return VerdictNeutral
	}

	parts := strings.Split(stripped, ",")
	firstUpper := strings.ToUpper(strings.TrimSpace(parts[0]))
	if sec == "rule" || firstUpper == "DOMAIN" || firstUpper == "DOMAIN-SUFFIX" || firstUpper == "DOMAIN-KEYWORD" || firstUpper == "DOMAIN-REGEX" || firstUpper == "DOMAIN-WILDCARD" || firstUpper == "URL-REGEX" || firstUpper == "IP-CIDR" || firstUpper == "IP-CIDR6" || firstUpper == "USER-AGENT" || firstUpper == "PROCESS-NAME" || firstUpper == "DEST-PORT" {
		policy := ""
		if len(parts) > 0 {
			policy = strings.ToUpper(strings.TrimSpace(parts[len(parts)-1]))
		}
		if REJECT_POLICIES[policy] {
			if containsAny(low, []string{"unlock", "解锁", "vip", "premium", "crack"}) {
				return VerdictEnhance
			}
			return VerdictAd
		}
		if len(parts) > 0 {
			if firstUpper == "DOMAIN" || firstUpper == "DOMAIN-SUFFIX" || firstUpper == "DOMAIN-KEYWORD" || firstUpper == "DOMAIN-REGEX" || firstUpper == "DOMAIN-WILDCARD" || firstUpper == "URL-REGEX" || firstUpper == "IP-CIDR" || firstUpper == "IP-CIDR6" || firstUpper == "USER-AGENT" || firstUpper == "PROCESS-NAME" || firstUpper == "DEST-PORT" {
				if containsAny(low, []string{"unlock", "解锁", "vip", "premium", "crack"}) {
					return VerdictEnhance
				}
				return VerdictAd
			}
		}
		return VerdictEnhance
	}

	if containsAny(low, ENHANCE_SCRIPT_MARKERS) {
		return VerdictEnhance
	}
	if containsAny(low, AD_SCRIPT_MARKERS) {
		return VerdictAd
	}
	return VerdictNeutral
}

func ShouldKeepPromaxLine(line string, section string, sourcePath string) bool {
	verdict := ClassifyPromaxLine(line, section, sourcePath)
	if verdict == VerdictAd {
		return true
	}
	if verdict == VerdictEnhance {
		return false
	}
	sec := strings.ToLower(strings.TrimSpace(section))
	if sec == "mitm" && strings.Contains(strings.ToLower(line), "hostname") {
		return true
	}
	return false
}

func ModuleIngestMode(path string, text string) string {
	name := strings.ToLower(filepath.Base(path))
	if strings.Contains(name, "解锁") || strings.Contains(name, "unlock") {
		if !containsAny(name, []string{"去广告", "adblock", "anti-ad"}) {
			return "skip"
		}
	}

	adLines := 0
	enhanceLines := 0
	currentSection := ""

	lines := strings.Split(text, "\n")
	for _, raw := range lines {
		stripped := strings.TrimSpace(raw)
		if strings.HasPrefix(stripped, "[") && strings.HasSuffix(stripped, "]") {
			currentSection = stripped[1 : len(stripped)-1]
			continue
		}
		if stripped == "" || strings.HasPrefix(stripped, "#") {
			continue
		}
		v := ClassifyPromaxLine(stripped, currentSection, path)
		if v == VerdictAd {
			adLines++
		} else if v == VerdictEnhance {
			enhanceLines++
		}
	}

	if adLines == 0 {
		return "skip"
	}
	if enhanceLines > 0 {
		return "split"
	}
	return "full"
}
